h2o-py/h2o/plot_result.py

```diff
@@ -0,0 +1,16 @@
+# -*- encoding: utf-8 -*-
+
+class H2OPlotResult(object):
+
+    def __init__(self, *args, figure=None):
+        self.figure = figure
+        self._tuple = args
+
+    def __iter__(self):
+        return iter(self._tuple)
+
+    def __getitem__(self, idx):
+        return self._tuple[idx]
+    
+    def tuple(self):
+        return self._tuple
```

Comment:
is this needed anywhere?
see my latest suggestion: https://github.com/h2oai/h2o-3/pull/5853#issuecomment-958028897

if we want to expose axes values explicitly on this container for example, the `*args` is here for convenience and backwards-compatibility, but I would not encourage users to call `result.tuple()`.
Instead we could decide to expose those with named params:
```python
def my_plot(…):
    xs = …
    ys = …
    fig = plt.figure()
    …
    return H2OPlotResult(xs, ys, figure=figure, xs=xs, ys=ys)
    
# consumer side:
xs, ys  = my_plot()  # backwards-compatible
fig = my_plot().figure  # consistent with all H2O plot functions
res = my_plot()
xs = res.xs    # explicit retrieval of x values
```
This way everything is accessible explicitly (as soon as documented in the `:return:` value for a given plot method).

Reply:
it's not. In tests I see it always extracted together as a tuple, so I would leave it without exposing explicit axes

Reply:
@koniecsveta you're right, I was wondering about the case where the user would want both the figure and the axes values…
but it appears that about anything can be then retrieved from the figure object, for example:
```python
xs = figure.axes[0].lines[0].get_xdata()
```
so yeah, I guess you don't need to expose the values explicitly then in this case.

There may other use-cases though, like `permutation_importance_plot`: user can't really retrieve the importance object from the figure, so we'll have to return sth like:
```python
def permutation_importance_plot(…):
    …
    return H2OPlotResult(importance, figure=figure, importance=importance) 
```
if we want the user to have access to both naturally, without doing some tricky assignments, then our current generic container doesn't work well :(
```python
imp = permutation_importance_plot(…)  # legacy code, however imp isn't the importance object anymore
imp, = permutation_importance_plot(…)  # this works with new impl though, but no one was using this because it didn't work with legacy code
imp = permutation_importance_plot(…).importance  # also works with new impl
```

in this case, the only way to get the old code working would be to use:
```python
def permutation_importance_plot(…):
    importance = …
    importance.figure = figure # making the assumption that we can extend the importance object
    return importance
```
it works, but problematic from a design perspective…

Reply:
yes I just hit on pdp plot , that's another problematic case

Reply:
Maybe I will first do design-perspective-problematic approach on all the functions that are problematic and then try to come up with some nicer solution when I will now what everything is problematic

Reply:
thinking about using a combination of mixin+descriptors for the `pdp/pip` use-cases, trying it and see if it can be nicely encapsulated

Reply:
what about?
```python
class HTuple(tuple): pass
class HList(list): pass
class HStr(str): pass

def get_decorated_plot_result(res, figure=None):
    # list all special/immutable types that we need first
    if res instanceof tuple:
       dec = HTuple(res)
    elif res instanceof list:
       dec = HList(res)
    elif res instanceof str:
       dec = HStr(res)
    else: # should be an H2O instance, should be mutable
       dec = res
    dec.figure = figure
    return dec
```
then in the plot fn:
```python
def my_plot(…):
   …
   return get_decorated_plot_result(old_result, figure=fig)
```

Reply:
updated the suggestion above after realizing that it doesn't handle the case where there was no return value (res=None)